### DISCUSSION TITLE
csp issue within wire-directives.js

### DISCUSSION THREAD

**Discussion by snapey** [livewire/livewire#4884] [1. Help]
I'm implementing CSP for a client, and using the Spatie package.

Adding csp_nonce() to the LivewireScripts blade tag has worked fine, but I have an error with wire-directives.js

https://github.com/livewire/livewire/blob/master/js/util/wire-directives.js#L93

This line throws a CSP error because 'unsafe-eval' is not an allowed source of script.

Any suggestions for a workaround.  I assume its related to `wire:` in the view, but there are many.

**Comment by joshhanley** [comment]
@snapey Livewire currently uses `eval()` in its codebase so it's possibly that, that is tripping it up. Plus the use of the function generated from a string in the code you linked.

I personally haven't worked with CSP, so can't provide a work around. Is there a way to whitelist the Livewire scripts?

**Comment by snapey** [comment]
I think it is only possible to apply  'unsafe-eval' for the entire site meaning that pentest tools will flag the CSP as being present but weak.

**Comment by liepumartins** [comment]
I also hope to see this changed!

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval#Never_use_eval

And CSP is hardly anything new: https://www.troyhunt.com/locking-down-your-website-scripts-with-csp-hashes-nonces-and-report-uri/

**Comment by binaryfire** [comment]
Is there a solution to this yet? Just started working with Livewire and had to allow `unsafe-eval` to get it working with CSP. Not ideal.

**Comment by piotrku** [comment]
Doesn't look good - over two years and still no solution.
Seems like Livewire is still not mature enough to be used for safe production environments. :(

**Reply by liepumartins** [reply]
I have stopped believing. :cry: :flying_saucer: 
All such systems I have come across relies on eval at some point, because avoiding it requires extra work both from the maintainers and the developers (end-users).

Until eval is removed from majority of browsers, I doubt there will be any change.

**Comment by kigathi-chege** [comment]
@snapey Did you find a workaround? I am facing the same issue.

**Reply by kigathi-chege** [reply]
I have had to do this:

```php
document.addEventListener('livewire:init', function() {
        document.getElementById('homeLoanCalcForm').addEventListener('submit', function(event) {
            event.preventDefault();
            Livewire.dispatch('calculate');
        });
    });
```

and then listen for the event inside my Livewire component like so:

#[On('calculate')]

**Comment by hari-salesdock** [comment]
@kigathi-chege @snapey 

I found that this error also occurs with Livewire's `wire:poll` functionality. Is there any workaround for this in Livewire version 3.5?

Also, can we always relay the JavaScript click event to perform Livewire actions (as you mentioned in your previous comment)?
